### ISSUE TITLE
Meeting freeze on unreliable connection

### ISSUE THREAD

**Issue by astas79** [aws/amazon-chime-sdk-ios#170]
**Describe the bug**
When there is a unreliable connection between 2 devices, one of the devices issues Dropped sample buffer error and local video freeze until meeting restart.

**To Reproduce**
Standard demo IOS app with serverless demo backend
IPhones Xs (iOS 14.2) + Ipad (iOS 14.2)
Connect to the same meeting on the 3G connection with video shared

**Expected behavior**
Meeting continues, instead local video on Iphone freeze until meeting restart. I was not able to reproduce on Ipad only happens on iPhone

**Logs**
xCode log is filling up with following messages:

2020-11-16 21:38:10.542367+0000 AmazonChimeSDKDemo[615:60226] [ERROR] MeetingModel - [013:087] [25891] (RTCLogging.mm:33): (RTCCameraVideoCapturer.m:311 -[RTCCameraVideoCapturer captureOutput:didDropSampleBuffer:fromConnection:]): Dropped sample buffer. Reason: OutOfBuffers
2020-11-16 21:38:10.608481+0000 AmazonChimeSDKDemo[615:60383] [ERROR] MeetingModel - [013:153] [68987] (RTCLogging.mm:33): (RTCCameraVideoCapturer.m:311 -[RTCCameraVideoCapturer captureOutput:didDropSampleBuffer:fromConnection:]): Dropped sample buffer. Reason: OutOfBuffers
2020-11-16 21:38:10.674829+0000 AmazonChimeSDKDemo[615:60383] [ERROR] MeetingModel - [013:219] [68987] (RTCLogging.mm:33): (RTCCameraVideoCapturer.m:311 -[RTCCameraVideoCapturer captureOutput:didDropSampleBuffer:fromConnection:]): Dropped sample buffer. Reason: OutOfBuffers
2020-11-16 21:38:10.743511+0000 AmazonChimeSDKDemo[615:60400] [ERROR] MeetingModel - [013:288]

**Test environment Info (please complete the following information):**
 - Device: [iPhoneXs]
 - OS: [iOS 14.2
 - Version AmazonChimeSDK: [e.g. 0.11.1]
 - Version AmazonChimeSDKMedia: [e.g. 0.7.1]

**Additional context**
[logs.txt](https://github.com/aws/amazon-chime-sdk-ios/files/5549950/logs.txt)

**Comment by alnlau** [comment]
Hello,

Thanks for reporting the issue and providing logs. We will take a look.

**Comment by alnlau** [comment]
Hello,

I was not able to reproduce this issue on my test device. Is the iPhone device that you're using doing anything else such as screen share / recording? Also, is this issue reproducible only on the iPhone Xs or on other iPhones as well? 

For the 3G connection, are you using Network Link Conditioner for the test? If so, what were the values for the settings?

**Comment by astas79** [comment]
Hi,

I was able to reproduce on 3 separate devices iPhone XS, XS Max and 11. Not reproducible on IPad. First noticed on IOS 14.2. And not happening on WiFi. 
I use "3G average" or "2G good" profiles in Xcode for network conditioning I believe caused by longer delays in packets delivery and also usually happens when there are a lot of changes in stream  - like rapid camera movements etc. When it freeze meeting cannot be restarted as the error comes up "unable to encode frame". Only complete shutdown of the app allows to restart. 
Iphones under normal load - no recording or sharing. CPU during the freeze about 20%.
Error messages stop when .stopLocalVideo is issued though starting local video brings back messages.

Hope that helps.

Thank you.

**Comment by tianyux-amz** [comment]
Hi @astas79 ,

`Dropped sample buffer. Reason: OutOfBuffers` means the sample buffer is full so we dropped the frames that cannot be processed in time.

For example, if the user is in an unstable network (2G/3G in your case), each frame will take longer to be sent. The sending side can drop to 5fps because of the low network bandwidth, but the capture rate won't be impacted by the network condition and will keep 15fps. In that case, 10 frames will be queued up in one second. Normally, the capture frame buffer size is 12, so user could run out of buffer in 1.2 second.

That being said, you should not see meeting freeze in this case. Instead, you may observe sluggish video and frame drops, and it should recover if the network condition gets improved.

I checked your logs, and seems like you ran into encoder failure:
```
2020-11-16 21:38:11.252431+0000 AmazonChimeSDKDemo[615:60183] [ERROR] MeetingModel - [013:797] [6155] (RTCVideoEncoderH264.mm:685): Failed to create compression session: -12912
2020-11-16 21:38:11.252496+0000 AmazonChimeSDKDemo[615:60183] [ERROR] MeetingModel - [013:797] [6155] (video_stream_encoder.cc:645): Failed to initialize the encoder associated with codec type: H264 (4)
2020-11-16 21:38:11.252531+0000 AmazonChimeSDKDemo[615:60183] [ERROR] MeetingModel - [013:797] [6155] (video_stream_encoder.cc:676): Failed to configure encoder.
2020-11-16 21:38:11.252611+0000 AmazonChimeSDKDemo[615:60183] [ERROR] MeetingModel - [013:797] [6155] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
2020-11-16 21:38:11.252786+0000 AmazonChimeSDKDemo[615:60183] [ERROR] MeetingModel - [013:797] [6155] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
```

Error code `-12912` is `kVTVideoEncoderMalfunctionErr`, which may happen when `VTSessionCopyProperty()` was called often, according to http://www.openradar.me/45889262.

I will forward this to our video team for further investigation. Will keep you posted.

**Comment by tianyux-amz** [comment]
Hi, FYI we have a potential fix for this issue, and it's expected to be out in the next release (scheduled on Dec 18).

**Comment by schonfeld** [comment]
> Hi, FYI we have a potential fix for this issue, and it's expected to be out in the next release (scheduled on Dec 18).

We've been seeing this happening *a lot* in production. Do you know when this fix might hit the `develop` branch?

**Comment by tianyux-amz** [comment]
The fix is in AmazonChimeSDKMedia.framework, which won't be available directly in the `development` branch.

Our QA team is still doing more verifications on the fixes. Will bring this up to the team and see if we can expedite the release.

**Comment by schonfeld** [comment]
> The fix is in AmazonChimeSDKMedia.framework, which won't be available directly in the `development` branch.
> 
> Our QA team is still doing more verifications on the fixes. Will bring this up to the team and see if we can expedite the release.

Thank you! What's the latest SDKMedia.framework that is unaffected by this bug? Do the {SDK, MediaSDK} have to be version paired? Or can we use the latest SDK with an older SDKMedia?

**Comment by tianyux-amz** [comment]
The potential fixes improve things to some extent, but QA reported that the issue was still reproducible. Our video team is still investigating the issue. Will keep you posted with the progress.

We use semantic versioning, so if the two Media libraries have the same minor version (e.g. v0.8.2 and v0.8.0), there should be no breaking change between them and they should be interchangeable.

**Comment by schonfeld** [comment]
We've spent a lot of dev time debugging this issue. Some findings & insights:

* This issue is not limited to poor connectivity. It can happen seemingly randomly with 2 stable connections. The encoder can complain of error code `-12912`, or `-12915`. See error excerpts below. The result seems identical, regardless of which error code is thrown.

* This seems to have started around early/mid-November. Date correlates with iOS 14.2 release. We've tested w/ iOS 14.3 RC, and this issue persists. Currently looking for <14.2 devices to test on, but they're very hard to find thanks to Apple's auto-updates. We've rolled back our app to a pre-November build, when we know for a fact this issue did not happen, and managed to reproduce even with the older build. Therefore, we've ruled out a code change issue on our end. 

* Our app is built using React Native, which makes our codebase virtually identical in Android. This issue does not happen in Android, at all. Android runs smoothly.




```
2020-12-09 10:48:50.322112-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:664] [139531] (RTCVideoEncoderH264.mm:685): Failed to create compression session: -12915
2020-12-09 10:48:50.322373-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:664] [139531] (RTCVideoEncoderH264.mm:685): Failed to create compression session: -12915
2020-12-09 10:48:50.323953-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:666] [139531] (video_stream_encoder.cc:645): Failed to initialize the encoder associated with codec type: H264 (4)
2020-12-09 10:48:50.324084-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:666] [139531] (video_stream_encoder.cc:645): Failed to initialize the encoder associated with codec type: H264 (4)
2020-12-09 10:48:50.324150-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:666] [139531] (video_stream_encoder.cc:676): Failed to configure encoder.
2020-12-09 10:48:50.324185-0500 Upstream Local[1675:180990] [ERROR] RNChime - [228:666] [139531] (video_stream_encoder.cc:676): Failed to configure encoder.
2020-12-09 10:48:51.145299-0500 Upstream Local[1675:180760] [ERROR] RNChime - [229:487] [130059] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
2020-12-09 10:48:51.145844-0500 Upstream Local[1675:180760] [ERROR] RNChime - [229:487] [130059] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
2020-12-09 10:48:51.228799-0500 Upstream Local[1675:180760] [ERROR] RNChime - [229:571] [130059] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
2020-12-09 10:48:51.229163-0500 Upstream Local[1675:180760] [ERROR] RNChime - [229:571] [130059] (video_stream_encoder.cc:1330): Failed to encode frame. Error code: -7
```

**Comment by schonfeld** [comment]
We've also tested this with the Amazon Chime iOS app (the _official_ one from the App Store), and while we can't access the logs to confirm, we've easily managed to reproduce the video freeze there as well.

**Comment by tianyux-amz** [comment]
Thanks for the details @schonfeld. We've also noticed that this issue seemed to be related to iOS 14.2, and the Media library is shared between Amazon Chime client and Amazon Chime SDK so both were impacted. I will pass your information to Video team as well. Thanks again for your help and patience!

**Comment by zhinang-amazon** [comment]
Hi @schonfeld 
We have a fix in 0.12.2 hotfix release that just went out.
Let us know if this is still an issue after using this new version.

**Comment by schonfeld** [comment]
> Hi @schonfeld
> We have a fix in 0.12.2 hotfix release that just went out.
> Let us know if this is still an issue after using this new version.

Thank you so much! We'll get this in the iOS review queue tonight. Will keep this thread posted if I see the issue again...

**Comment by alnlau** [comment]
Hello @schonfeld,

Are you still encountering the issue after upgrading?
